### DISCUSSION TITLE
Using beartype with PyTorch

### DISCUSSION THREAD

**Discussion by amitkparekh** [beartype/beartype#96] [Q&A] [answered]
I love the idea of this project and I _want_ to use beartype in all my projects going forwards. I've been lurking for a while and I don't think it's been suggested so here's my question/request: 

How can I use this with [torch's Tensor](https://pytorch.org/docs/stable/tensors.html)? 

I'm writing a lot of ML/NLP research code and I want to use this in my projects to not need to keep typing the shape in the comments, and so I can just further trust that each tensor is what I've told it to be. 

I know you support numpy OOTB, but I'm not sure whether it can translate directly to torch, and whether it can follow the various reshaping methods that make square tensors fit in round-shaped holes. 

I've had trouble fully grasping how to do things with beartype because I have trouble understanding the README. Don't get me wrong, I absolutely _love_ the writing, but I genuinely have trouble understanding it because it's just a lot of information that adds more information super fast.

**Comment by leycec** [comment, marked as answer]
> ...because I have trouble understanding the README.

**You are not alone.** We also have trouble understanding 1,613-page doctoral theses in obscure LaTeX3.1415 formats littered with unpronounceable acronyms, jargon-on-jargon, and recursive puns – which is why I let my wife read those. In this case, beartype usage could best be synopsized with this fist-pumping motto:

* **[`beartype.vale.Is[lambda obj: ...]`](https://github.com/beartype/beartype#validator-api) is QA King.**

[`beartype.vale.Is[lambda obj: ...]`](https://github.com/beartype/beartype#validator-api) is our functional validator API. It's what you do when all of the following apply:

* **What you want to do isn't covered by an existing PEP standard.** Almost certain that's the case here.
* **You don't want to dive *too* deep into beartype validators.** Efficient alternatives to our functional validator API do exist – but they're *much* less general-purpose. For most purposes (almost certainly including this one), the functional validator API suffices. Always go there first.
* **You know exactly how, what, and why you want to validate.** In this case, "How can I use this with [torch's Tensor](https://pytorch.org/docs/stable/tensors.html)?" is a *really* broad question. To preserve dwindling sanity, let's pretend you instead asked:

> "How can I type-check a PyTorch tensor to have the expected dtype?"

Apropos question, Amit. All of the above conditions now apply! We can now type you up a possibly working minimal-length example (MLE) combining the fearsome powers of @beartype **+** PyTorch **+** [PEP 593](https://www.python.org/dev/peps/pep-0593) to validate callable parameters and returns as strictly floating-point *or* integral PyTorch tensors:

``` python
from beartype import beartype
from beartype.vale import Is
from torch import (
    float as torch_float,
    int as torch_int,
    tensor,
)
from typing import Annotated   # <--------------- if Python ≥ 3.9.0
# from typing_extensions import Annotated   # <-- if Python < 3.9.0

# PEP-compliant type hint matching only a floating-point PyTorch tensor.
TorchTensorFloat = Annotated[tensor, Is[
    lambda tens: tens.type() is torch_float]]

# PEP-compliant type hint matching only an integral PyTorch tensor.
TorchTensorInt = Annotated[tensor, Is[
    lambda tens: tens.type() is torch_int]]

# Type-check everything like an NLP babelfish.
@beartype
def deep_dream(dreamy_tensor: TorchTensorFloat) -> TorchTensorInt:
    return dreamy_tensor.type(dtype=torch_int)
```

Of course, `beartype.vale.Is[...]` supports arbitrarily complex Turing-complete Python expressions. The above MLE generalizes to typing the dimensionality, device, and other metadata of a tensor to whatever degree of specificity you need.

`beartype.vale.Is[...]`: *it's lambdas all the way down.*

**Reply by amitkparekh** [reply]
Thank you so much for a brilliant answer with the MLE to suit! I didn't know if beartype supported torch because the README didn't mention it — or I didn't see that it mentioned it, which is also quite possible. 

One more question though if that's okay, if you're not using the `@beartype` decorator, does it still need to be imported? 

That said, I'm going to start adding beartype verification (or _bear-ification_ _<sup>I'll show myself out...</sup>_) to all my ML code from now own.

**Comment by leycec** [comment]
> I didn't know if beartype supported torch because the README didn't mention it...

**You're absolutely right.** Technically, @beartype *indirectly* supports everything (including PyTorch) for fanatical end users willing to sacrifice their weekends and wives by writing their own `beartype.vale.Is[...]`-fueled type hints at home. Pragmatically, our documentation *really* needs to highlight that @beartype trivializes PyTorch tensor typing – and how exactly to make that happen.

As a self-reminder, I've transitioned your excellent questionnaire into a full-fledged feature request on our issue tracker. Prod me if I conveniently forget to do this. :probing_cane: 

> ...if you're not using the `@beartype` decorator, does it still need to be imported?

**You're absolutely right,** yet again. You're two for two here! I'd buy five lotto tickets with those odds. <sup>Suggested numbers: *4, 8, 15, 16, 23, 42.*</sup>

I accidentally dropped the `@beartype` decorator above, but just patched that back in. You can of course omit `@beartype` – but then you're not actually type-checking anything, which kinda defeats the point. Cue sad yet smiling emoji. :smiling_face_with_tear: 

> *bear-ification* <sup>*I'll show myself out...*</sup>

![heh](https://i2.wp.com/punpedia.org/wp-content/uploads/2016/09/Igotnothing_163cbc50a8a0bd1b08e71625cba3f58f.png?ssl=1)

**Comment by JulesGM** [comment]
Wanted to say @leycec that the NumPy typing support was really awesome.
There doesn't seem to be as in-depth of an annotation system in Pytorch?

**Comment by leycec** [comment]
There's a great question. Surprisingly, I even have a half-great answer for you. 

**You are absolutely correct.** Unlike NumPy, PyTorch never received the interoffice memo we circulated last year that:

> Untyped mission-critical data structures is a bad look in 2022.

Like NumPy, PyTorch *has* received a bevy of unofficial third-party packages purporting to type PyTorch tensors. The most popular appears to be [`torchtyping`](https://github.com/patrick-kidger/torchtyping). Because compsci, hard tradeoffs exist here.

On the one hand, `torchtyping` is PEP-noncompliant. That's bad, because it means PEP-aware IDEs like PyCharm and VSCode will vomit all over your code. On the other hand, `torchtyping` supplies a mypy plugin. That's good, because at least mypy will be willing to look the other way. On the gripping hand, all transitive downstream dependencies of your project will need to manually enable that mypy plugin. That's bad, because you don't control those dependencies.

Let's start a new pair of hands. The old ones are already cramping up on me. On the new hand, `torchtyping` fails to support @beartype or other actively maintained runtime type checkers. That's because, because we are @beartype. On the other new hand, `torchtyping` provides a risky `patch_typeguard()` function dynamically monkey-patching `typeguard` to support its PEP-noncompliant type hints. That's good, because only life on the precipice of disaster is life worth living. On the new gripping hand, [**`typeguard` is no longer actively maintained**](https://github.com/agronholm/typeguard/issues/198). And that's as bad as it gets in this industry, really.

## tl;dr

`torchtyping` violates PEP standards *and* fails to support anything meaningful except mypy. You probably shouldn't use it unless PyTorch officially adopts, embraces, and extends it into the official `pytorch` package. Your mileage may vary, because my speedometer blatantly falsifies data.

Incidentally, the grumpiest angel I've ever seen:  :angel:

I can smell the smug from here, folks. It's like the [Van Halen *1984* cover cherubim](https://www.coverourtracks.com/single-post/2016/06/27/margo-nahas-on-her-iconic-cover-for-van-halens-1984) got an exhausting day job in Redmond and just went downhill from there.

**Comment by JulesGM** [comment]
I'll write an issue on the Pytorch Github I guess.

**Comment by leycec** [comment]
**Frabjous idea.** They're prolly *not* gonna be receptive – but they should be. The hard blocker for them is that refactoring their hints to be PEP-compliant means:

* **Deprecating their entire existing API.** What they have fundamentally violates typing standards. They've gotta chuck that out with the tepid bathwater I just bathed the cat in. No one wants to destroy their API. Sometimes, one must destroy their API to built a better future for others.
* **Declaring a new API fuelled by [PEP 593](https://www.python.org/dev/peps/pep-0593).** This API should resemble:

```python
from torch import Tensor
from torchtyping import TensorType
from typing import Annotated

# PEP-compliant type hint matching only batch floating-point
# PyTorch tensors.
TensorBatchFloat = Annotated[Tensor, TensorType['batch', float]]
```

:muscle: :bear:

**Comment by patrick-kidger** [comment]
Just chiming in here as the author of TorchTyping!
(I found this thread whilst procrastinating and waiting for some tests to pass...)

TL;DR: PEP 646 now exists! I'm receptive to chucking out the cat-scented bathwater. Let's figure something out.

---

Right, first -- in defense of the design decisions for TorchTyping. These were carefully, and conciously, chosen.

PEP violations. TorchTyping was written at a time when variadic generics did not exist, and PEP 646 was still stuck in some kind of development hell. What TorchTyping had instead was a facsimile jury-rigged out of spare parts and bits of string, and then dressed up to look pretty. So PEP compatibility was clearly never on the table. Mypy compatibility was right out. Why do you think TorchTyping ended up using a runtime type checker? ;)

With that decision made, TorchTyping ended up being loads of fun. Because now we can start doing stuff like `TensorType["batch": ..., "channels"]` and end up with readable type annotations that people actually want to use. And -- without any of the usual unreadable `T = TypeVar('T')` nonsense that so pollutes the rest of the Python typing ecosystem.

---

However, we now live in the future. PEP 646 exists, and we have been graced with an official way to make such things work. We should do something about this. (And if necessary hold our nose when the syntax is no longer as pretty.)

My strong belief is that the best thing to do now is to write a brand-new package for representing Array/Tensor types. Pretty much what the discussion in PEP 646 invites us to do, in fact.

In particular I'm imagining something that works for all NumPy/PyTorch/JAX/whatever the latest arrray-based widget is. Because honestly who has time to keep reimplementing the same thing? I'm using JAX these days, dammit, I want a library that works with that too. Something with pleasing syntax akin (but definitely not exactly) like TorchTyping. Something that is PEP compliant. And of course, crucially, something with beartype support.

I'm very familiar with what's expected for array manipulation (including multiple notions of variadic-ness that can interact in fun ways, I can spin you a yarn there). @leycec you're obviously key to making sure some new endeavour stays on the correct side of PEP compliance, and making beartype do its magic. So if you're interested, I'm interested -- let's make this happen?

**Comment by leycec** [comment]
**Let's type this.**

> In particular I'm imagining something that works for all NumPy/PyTorch/JAX/whatever the latest arrray-based widget is.

So much this, which is why I was visibly dragging my Achilles heels here. :sweat: 

I'm currently [putting out fires all over the codebase](https://github.com/beartype/beartype/issues/104) and on the fine precipice of physical collapse. <sup>*it's the knife's edge, bro*</sup> So, I'd just like to publicly announce that **I was shocked, simply shocked** to learn from `torchtyping` that `[...]`-style subscription syntax implicitly supports keyword arguments via dictionary syntax:

```python
# Literally cannot believe this parses... but it does. Oh, how it does.
from torchtyping import TensorType
TensorType["dim1": ..., "dim2": ...]
```

Still blows my mind to this day and I only learned about that yesterday. Thanks to that inscrutable magic, we can actually do everything you and I and everyone else want. I am now thinking of a new third-party API that rhymes with:

```
# Import generic tensor annotation machinery.
from tensorflex import FlexyTensor
from typing import Annotated   # <--------------- if Python ≥ 3.9.0
# from typing_extensions import Annotated   # <-- if Python < 3.9.0

# Generic object typing any two-dimensional floating-point tensor.
# Since this is *not* PEP-compliant in and of itself, this is intended
# to be embedded in a PEP-compliant type hint of the form:
#    typing.Annotated[{tensor_type}, FlexyTensor[...]]
TensorFloat2D = FlexyTensor['dtype': float, 'dim': 2]

# ..................{ NUMPY                      }..................
# Annotate NumPy arrays with aplomb!
from numpy import ndarray

# PEP-compliant type hint matching any two-dimensional
# floating-point NumPy array. \o/
NDArrayFloat2D = Annotated[ndarray, TensorFloat2D]

# ..................{ PYTORCH                      }..................
# Annotate PyTorch arrays with equal aplomb!
from torch import Tensor

# PEP-compliant type hint matching any two-dimensional
# floating-point PyTorch tensor. \o/ \o/
TensorFloat2D = Annotated[Tensor, TensorFloat2D]
```

![galaxy brain explodes](https://c.tenor.com/bD9vHNiR1rQAAAAS/boom-mind-blown.gif)

**Comment by patrick-kidger** [comment]
Haha yup, the `key: value` syntax ends up being really useful for annotating named groups of dimensions. (It's also used in a bunch of other edge cases.)

[
Side note:
The use of key-value syntax is really just a neat trick.
- For any `Annotated`-based approach then you can put pretty much arbitrary stuff on the right hand side so just go ahead and write arbitrary Python.
- For any mypy-compatible approach then... mypy doesn't support any kind of `key: value` syntax, at least that I know of.
]

In terms of making a new library work:
- Option 1: use `Annotated`. (Your suggestion.)
    - Problem 1: It's way too much to have typed-out as an end-user.
        - One tends to write a lot of functions. I'm not writing something stupidly verbose like
           ```python
           TensorBatchXChannels = Annotated[Tensor, TensorType["batch", "x_channels"]]
           TensorBatchYChannels = Annotated[Tensor, TensorType["batch", "y_channels"]]
           TensorBatchXYChannels = Annotated[Tensor, TensorType["batch", "x_channels", "y_channels"]]

           def batch_outer_product(x: TensorBatchXChannels, y: TensorBatchYChannels) -> TensorBatchXYChannels:
               return x[:, :, None] * y[:, None, :]
           ```
           and neither is anyone else.
        - Seriously, there's been other array/tensor-typing libraries before. TorchTyping is probably more popular than all of them put together. Why? Syntax that is actually usable.
    - Problem 2: Can something like ever work with Mypy? Do we need to write an extension? Who knows. By default it all gets ignored. :(
    - (Problem 3: It doesn't use variadic generics. We have a shiny new toy and I want to ~~break it~~ use it.)
    - *Verdict: not suitable.*
- Option 2: wrap `Annotated`. (TorchTyping.)
    - We can fix "Problem 1", at least, by defining some neater syntax ourselves. Let's override `__class_getitem__` and make it lower `TensorType[...]` into `Annotated[Tensor, {...}]`. **And this is exactly what TorchTyping already does.**
    - But it doesn't fix "Problem 2". It's still sitting outside the formal specification of the Python type system.
    - *Verdict: current status quo. Can we do better?*
- Option 3: Generics and/or Protocols.
    - Hurrah, we have variadic generics now. Let's do something nice with these.
    - **Using these are what I'm suggesting.**
    - *Verdict: I don't know -- need to try this.*
- Option 4: Add mypy/beartype support to Option 2.
    - Adding beartype support to Option 2 would be essentially straightforward; just have beartype introspect the `Annotated` object, exactly as `typeguard` (via `patch_typeguard`) already does.
    - Adding mypy support isn't something I've really looked at (as a non-user myself), but it may be possible to write a plugin.
    - *Verdict: backup plan?*

WDYT? I've been drafting some stuff in the vein of Option 3 and I think it looks plausible.

**Comment by JulesGM** [comment]
Created an issue on the Pytorch repos. https://github.com/pytorch/pytorch/issues/73359
A member replied, you folks may be interested. 
@leycec @patrick-kidger

**Comment by rsokl** [comment]
> Option 3: Generics and/or Protocols.

Hi @leycec, @patrick-kidger, @JulesGM ! I've been thinking about this problem some, and have an interesting solution along the lines of Option 3: https://github.com/rsokl/phantom-tensors

High-level features:
- Leverages PEP 646: supports variadic generics to express flexible type information
- Compatible with static type checkers, no mypy plugin needed
- Ergonomic and powerful parsing: validate (at runtime) tensor shape consistency and carry proof-of-validation forward via type-narrowing (statically)
- Works out of the box with beartype and other runtime type checkers -- including shape-consistency checks across tensors
- Library agnostic: have prototypes working for torch and numpy. At this stage it is trivial to add support for other array-based libraries

Note that this is a super rough prototype. I've only put < 10 hours into this. Exploring beyond what is shown in the readme will surely lead to bugs/crashes.

Would love to get your thoughts!

**Reply by patrick-kidger** [reply]
Hey there! A while back I gave static type-checking of array types a try, and gave it up as a bad job. So this is cool to see!

As requested, here's some thoughts.

1. Heads-up that you're missing a `py.typed` file.

2. It looks like making this work is going to require adding type stubs for the stdlib NumPy/PyTorch(/JAX/....)? This was always one of the big hurdles for me -- I don't have time to create these / maintain these, and the library maintainers generally haven't seemed too fussed about doing so either.

3. Doesn't making `phantom_tensors.numpy.NDArray` a subclass of `numpy.typing.NDArray` introduce some kind of static-type checking issues? Since you're not actually passing these subclasses around, you can usually cook up examples where static-type checkers complain.

4. Does this work with IDEs? (I.e. can they figure out what the type is and offer method hints?) This was always one of the big complaints about TorchTyping. (And something I think jaxtyping probably fixes, although I've not verified that.)

Other comments on static type-checking of array types in general:

1. From an end-user perspective, all of the `NewType`s and `TypeVar`s floating around were always pretty impenetrable. This is really Python's fault for such awful syntax, but making this actually usable for your boots-on-the-ground data-scientist was again part of what drove me away from static type checking. Any thoughts?

2. Likewise the need for `cast`/`parse` to lift value-level axes into type-level axes is frustrating, e.g. every time you use a `concatenate` operation -- which is outside the realm of static type checking as it would need shape arithmetic to express. And `concatenate` is a pretty common operation.

3. I have a bit of a rant about static+array-type-checking [here](https://github.com/patrick-kidger/torchtyping/issues/37#issuecomment-1153294196). I conclude by saying that I think it should be possible to exert a lot of effort in order to get static type-checking of shapes... but not dtypes, integer axes, or broadcasting, and that doing so would requiring creating+maintaining type stubs for all of NumPy/PyTorch/... . Do I understand correctly that that's pretty much what you're aiming to do?

**Reply by rsokl** [reply]
Thanks for sharing your thoughts!

> It looks like making this work is going to require adding type stubs for the stdlib NumPy/PyTorch(/JAX/....)? This was always one of the big hurdles for me [...] creating+maintaining type stubs for all of NumPy/PyTorch/... . Do I understand correctly that that's pretty much what you're aiming to do?

Goodness gracious *no*. Not even if I was paid to. That battle is lost before it even begins!

`phantom-tensors` is not meant to provide shape-type information for general/all numpy/pytorch/jax/... operations. That's just not doable without things like shape arithmetic. Rather, I want to be able to write interfaces like:

```python
def positional_embedding(x: Tensor[B, T]) -> Tensor[B, T, C]: ...

def multihead_attn(x: Tensor[B, Tq, C], mask: Tensor[T, T]) -> Tensor[B, Tq, C]: ...

def layer_norm(x: Tensor[..., C]) -> Tensor[..., C]: ...

def decoder_layer(x: Tensor[B, T, C], mask: Tensor[T, T]) -> Tensor[B, T, C]: ... 
```
and to treat those annotations as *verified documentation*. Verified both statically and at runtime (and regardless of the tensor library I am using). Thus phantom-tensors are about being able to express these shapes, get accurate static type checking for them, and having convenient ways to "parse" my tensors so that I can "prove" – at runtime – that their shapes are consistent with their annotations (and consistent with the annotations of accompanying tensors).

So, I am totally okay with the fact that my tensor's type will widen from `Tensor[B, T]` to `Tensor` the moment I do math on it within `positional_embedding`. I just want validation at the *interfaces*. This will catch 80%+ of issues that are associated with shape stuff. The tools it provides can also be really useful for doing intermediate debugging within the function too - just add some extra `parse` calls.

I know the current README makes it look like one is just parsing all over the place and constantly doing casting, but that is mainly because the docs are rapid-firing quick examples to the reader. The intended use case would concentrate all of the parsing towards the top of the program and at the edges of the interfaces, and everything else would (hopefully) just.. vibe.

> Doesn't making phantom_tensors.numpy.NDArray a subclass of numpy.typing.NDArray introduce some kind of static-type checking issues?

So far I haven't seen any issues. I can pass a phantom_tensors ndarray to native numpy functions without any false positives from the type checker.

```python
x = parse(np.ones((2, 3)), NDArray[A, B])
out = np.multiply(x, x)  # type checker sees: NDArray[Any]  -- the type widens back out as-expected
``` 

NumPy is tricky because `ndarray` is already a generic. The current state of things was just me hacking around with it last night just to prove to myself that I could get parity with `Tensor`. Eventually I would need to think about handling things like dtype (and maybe device for torch). But that all seems to be relatively straight forward (but potentially really annoying) compared to proving out the shape aspect of things.

> Does this work with IDEs?

Yep! Working in IDEs basically comes down to working with static type checkers.

<img src="https://user-images.githubusercontent.com/29104956/190538664-80c5d98b-21e4-40db-b99e-c36cc8a6c707.png" width="500" />

<img src="https://user-images.githubusercontent.com/29104956/190538955-6706eaf1-9698-40b9-96f5-d0a5bcc6b3c4.png" width="400" />

> From an end-user perspective, all of the NewTypes and TypeVars floating around were always pretty impenetrable.

Totally agree. And I think that this library could do some things to make some of this quite a bit more ergonomic. Something as simple as:

```python
from phantom_tensors.letters import A, B, C
from phantom_tensors.words import Height, Width, Channel
```

where these are simply `NewType(<name>, int)` that have been defined for the user. This could really reduce the boilerplate for playing around with the library. But if you look at the signatures for the GPT example above, there isn't *too* much that needs to be done in terms of declaring these.

Additionally, I could imagine providing library-specific overrides *only* for common tensor creation and conversion functions, so that it is easier to create tensors with the correct shape-types. E.g., something like the following would be nice:

```python 
from phantom_tensors.torch import ones

x = parse(ones((2, 3)), Tensor[A, B])

a, b = x.shape  # type checker sees: A, B  (this works already)

y = ones((b, a, b, a))
reveal_type(y)  # *will* reveal: Tensor[B, A, B, A] (doesn't work yet)
```

This is where I think doing some library-specific overrides could make these types easier to propagate.

> I have a bit of a rant about static+array-type-checking https://github.com/patrick-kidger/torchtyping/issues/37#issuecomment-1153294196.

Yeah, I totally get where you are coming from with this[^1]. I hope that I have conveyed now that I mean to come at this from a different angle, where I am trying to be judicious and get the biggest bang for my buck while using the most light(-and-PEP-compliant-) touch that I can manage. Other attempts that I have seen have been more all-or-nothing than I think is necessary.

I am excited mainly because it seems like there *is* a light at the end of this particular tunnel, and that, whereas the all-in approach would be an infinity of work for a single library, this might actually scale to *most* relevant libraries pretty easily.

[^1]: But I don't think you should give up on static type checking! It is *really* useful having tool-assisted error checking in real time, and to be able to catch bugs in static code.

**Reply by patrick-kidger** [reply]
> Goodness gracious no. Not even if I was paid to. That battle is lost before it even begins!

Aha, that's good to hear. It would take a brave soul to take that on. :D

One question I do have though. How is this:

> So, I am totally okay with the fact that my tensor's type will widen from `Tensor[B, T]` to `Tensor`

compatible with this:

> I just want validation at the interfaces.

?

1. If you mean static type checking then this doesn't make sense to me.

2. If you mean runtime type checking then this is what jaxtyping/torchtyping already do, with greater flexibility in what they can express. (And the former at least has some partial support for static type checking, and [there are plans to backport this support to torchtyping](https://github.com/google/jaxtyping/issues/18).)

> Doesn't making phantom_tensors.numpy.NDArray a subclass of numpy.typing.NDArray introduce some kind of static-type checking issues?

What about with return types? Or  I did do a quick check with this and couldn't quite reproduce what I had in mind, but if you do

```python
def foo(...) -> DerivedClass:
    return BaseClass()
```

then I would expect a typechecker to throw an error. (And we're in this case as your type annotation is a derived class of the true type?)

> dtype (and maybe device for torch)

I think this is super tricky to have a static syntax for, since there's no notion of optionality. For example if you support dtypes then you have to write `Array[...shape..., Any]` to accept any dtype, and not merely `Array[...shape...]`. With device as well then all of your shape-only annotations necessarily become `Array[...shape..., Any, Any]`.

(Alternatively a trick like jaxtyping's may work? e.g. `Float[Array, ...shape...]`)

> Additionally, I could imagine providing library-specific overrides only for common tensor creation and conversion functions

FWIW I think I'd recommend against this. At least for me personally, this is the point at which the library overhead (/API surface) dominates the library benefit. (Especially if you want to use some other library that also needs the same thing, [e.g. to support units](https://github.com/dfm/jpu).)

> But I don't think you should give up on static type checking!

FWIW, since I wrote that: jaxtyping does support static type checking. Partially. It doesn't handle shape/dtype checks, but it does at least check the array/tensor-ness of the type. Progress!

---

I've picked out the things I felt worth commenting on -- otherwise I agree with what you're saying, and I agree it's exciting to see some possible light at the end of the tunnel. (Even if at this point I don't completely understand how you seem to be getting there :D )
